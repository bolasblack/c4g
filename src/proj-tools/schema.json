{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsSchematicSchema",
  "title": "Schematic Options Schema",
  "definitions": {
    "includeItem": {
      "$id": "#includeItem",
      "type": "string",
      "enum": [
        "lint-staged",
        "commitlint",
        "githooks",
        "renovate",
        "jest",
        "prettier"
      ]
    }
  },
  "type": "object",
  "properties": {
    "include": {
      "type": "array",
      "items": {
        "$ref": "#includeItem"
      },
      "description": "The project tools want add"
    },
    "cwd": {
      "type": "string",
      "description": "Specify the schematic execute cwd relative to current cwd"
    },
    "interactive": {
      "type": "boolean",
      "default": true,
      "description": "Ask some option interactively"
    },
    "jest-react": {
      "type": "boolean",
      "default": true,
      "description": "Add react support to jest"
    }
  },
  "required": []
}
