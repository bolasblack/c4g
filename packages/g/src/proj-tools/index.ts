import { strings } from '@angular-devkit/core'
import {
  Rule,
  mergeWith,
  chain,
  callRule,
  apply,
  move,
  template,
} from '@angular-devkit/schematics'
import { Observable, of, empty, from } from 'rxjs'
import { mergeMap, toArray } from 'rxjs/operators'
import * as inquirer from 'inquirer'
import * as path from 'path'
import { produce } from 'immer'
import { file as fileSource } from '@c4605/schematic-utils/lib/sources/file'
import {
  installNodePackage,
  NodePackageType,
} from '@c4605/schematic-utils/lib/rules/installNodePackage'
import { file as fileRule } from '@c4605/schematic-utils/lib/rules/file'
import { when } from '@c4605/schematic-utils/lib/rules/when'
import { Schema as Options, IncludeItem } from './schema'

export { featuresEnabled } from './featureEnabled'

export { Options, IncludeItem }

export function main(options: Options): Rule {
  return (tree, ctx) =>
    (options.interactive ? requestOptions(options) : of(options)).pipe(
      mergeMap((opts) => {
        opts.include = opts.include || []
        return from(opts.include)
      }),
      mergeMap(getRule.bind(null, options)),
      toArray(),
      mergeMap((rules) => {
        if (rules.length) {
          rules = rules.concat(
            installNodePackage({
              packageNames: ['@c4605/toolconfs'],
              workingDirectory: options.cwd,
            }),
          )
        }

        return callRule(chain(rules), tree, ctx)
      }),
    )
}

function getRule(options: Options, includeItem: IncludeItem): Observable<Rule> {
  const install = (opts: Parameters<typeof installNodePackage>[0]) =>
    installNodePackage({
      workingDirectory: options.cwd,
      type: NodePackageType.Dev,
      ...opts,
    })

  const filePath = (relativePath: string) =>
    path.join(options.cwd || '', relativePath)

  switch (includeItem) {
    case IncludeItem.Prettier:
      return of(
        chain([
          install({
            packageNames: ['prettier'],
          }),
          fileRule(filePath('package.json'), (content: object) => ({
            ...content,
            prettier: '@c4605/toolconfs/prettierrc',
          })),
        ]),
      )

    case IncludeItem.Commitlint:
      return of(
        chain([
          install({
            packageNames: ['commitlint', '@commitlint/config-conventional'],
          }),
          configFile(options, '.commitlintrc.js'),
        ]),
      )

    case IncludeItem.LintStaged:
      return of(
        chain([
          install({ packageNames: ['lint-staged'] }),
          configFile(options, 'lint-staged.config.js'),
        ]),
      )

    case IncludeItem.Githooks:
      return of(install({ packageNames: ['git-hook-pure'] }))

    case IncludeItem.Renovate:
      return of(configFile(options, 'renovate.json'))

    case IncludeItem.Eslint:
      return of(eslintRule(options))

    case IncludeItem.Jest:
      return of(jestRule(options))
  }

  return empty()
}

function configFile(options: Options, name: string, rules: Rule[] = []): Rule {
  return mergeWith(
    apply(fileSource(`./files/${name}`), [move(options.cwd || ''), ...rules]),
  )
}

export function eslintRule(options: Options): Rule {
  return chain([
    configFile(options, '.eslintrc.json'),
    when(
      () => options.include!.includes(IncludeItem.Prettier),
      fileRule(
        path.join(options.cwd || '', '.eslintrc.json'),
        (content: Record<string, any>) => ({
          ...content,
          extends: [
            ...content.extends,
            './node_modules/@c4605/toolconfs/eslintrc.prettier',
          ],
        }),
      ),
    ),
    installNodePackage({
      packageNames: [
        'eslint',
        options.include!.includes(IncludeItem.Prettier)
          ? 'eslint-config-prettier'
          : null,
        '@typescript-eslint/parser',
        '@typescript-eslint/eslint-plugin',
      ].filter((s): s is string => Boolean(s)),
      workingDirectory: options.cwd,
      type: NodePackageType.Dev,
    }),
  ])
}

export function jestRule(options: Options = {}): Rule {
  const pkgJsonPath = path.join(options.cwd || '', 'package.json')

  const completedOpts: Required<Options> = Object.assign(
    {
      cwd: '',
      include: [],
      interactive: false,
      jestReact: false,
    },
    options,
  )
  const templateOpts = {
    ...completedOpts,
    dasherize: strings.dasherize,
  }

  return chain([
    addJestDependencies(options),
    configFile(options, 'jest.config.js', [template(templateOpts)]),
    when(
      (tree) => tree.exists(pkgJsonPath),
      fileRule(
        pkgJsonPath,
        produce((content: any) => {
          content.scripts = content.scripts || {}
          if (content.scripts.test) {
            content.scripts.test += '&& jest ./src'
          } else {
            content.scripts.test = 'jest ./src'
          }
        }),
      ),
    ),
  ])

  function addJestDependencies(options: Options): Rule {
    const { jestReact } = options

    const deps = ['jest', 'jest-cli', 'ts-jest', '@types/jest']

    if (jestReact) {
      deps.push(
        'enzyme',
        'enzyme-to-json',
        'enzyme-adapter-react-16',
        'identity-obj-proxy',
      )
    }

    return installNodePackage({
      type: NodePackageType.Dev,
      packageNames: deps,
      workingDirectory: options.cwd,
    })
  }
}

export const requestOptions = (
  opts: Partial<Options> = {},
): Observable<Options> =>
  from(
    inquirer.prompt([
      {
        name: 'include',
        type: 'checkbox',
        choices: Object.values(IncludeItem),
        default: opts.include || [],
      },
      {
        name: 'jestReact',
        type: 'confirm',
        message: 'jest add react support?',
        default: false,
        when(ans: Options) {
          return (
            opts.jestReact == null && ans.include!.includes(IncludeItem.Jest)
          )
        },
      },
    ]),
  )
