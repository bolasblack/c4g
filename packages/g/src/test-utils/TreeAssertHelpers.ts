import {
  UnitTestTree,
  SchematicTestRunner,
} from '@angular-devkit/schematics/testing'

export function matchFileStructureSnapshot(
  creator: () => Promise<{ tree: UnitTestTree }>,
) {
  it('match file structure snapshot', async () => {
    const { tree } = await creator()
    expect(tree.files).toMatchSnapshot('file structure')
  })
}

export function matchFileContentSnapshot(
  creator: () => Promise<{ tree: UnitTestTree }>,
) {
  it('match file content snapshot', async () => {
    const { tree } = await creator()
    tree.files.forEach(file => {
      expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
    })
  })
}

export function matchTasksSnapshot(
  creator: () => Promise<{ runner: SchematicTestRunner }>,
) {
  it('match tasks snapshot', async () => {
    const { runner } = await creator()
    expect(runner.tasks).toMatchSnapshot('rule tasks')
  })
}

export function matchRunResult(
  creator: () => Promise<{
    tree: UnitTestTree
    runner: SchematicTestRunner
  }>,
) {
  matchFileStructureSnapshot(creator)
  matchFileContentSnapshot(creator)
  matchTasksSnapshot(creator)
}

export function assertTreeSnapshot(
  runner: SchematicTestRunner,
  tree: UnitTestTree,
) {
  expect(tree.files).toMatchSnapshot('file structure')

  tree.files.forEach(file => {
    expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
  })

  expect(runner.tasks).toMatchSnapshot('rule tasks')
}
