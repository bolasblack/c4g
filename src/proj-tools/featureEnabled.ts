import { IncludeItem } from './schema'
import { TypedSchematicContext, DirEntry } from '@angular-devkit/schematics'

export interface FeaturesEnabledTypedSchematicContext
  extends TypedSchematicContext<{}, {}> {
  enabledFeatures?: IncludeItem[]
}

export function featuresEnabled(
  features: IncludeItem[],
): (
  dir: DirEntry,
  ctx: FeaturesEnabledTypedSchematicContext | TypedSchematicContext<{}, {}>,
) => boolean {
  return (tree, ctx) => features.every(f => singleFeatureEnabled(f, tree, ctx))
}

function singleFeatureEnabled(
  feature: IncludeItem,
  dir: DirEntry,
  ctx: FeaturesEnabledTypedSchematicContext | TypedSchematicContext<{}, {}>,
): boolean {
  if (
    process.env.NODE_ENV === 'test' &&
    'enabledFeatures' in ctx &&
    ctx.enabledFeatures
  ) {
    return ctx.enabledFeatures.includes(feature)
  }

  switch (feature) {
    case IncludeItem.Commitlint:
      return commitlintEnabled(dir, ctx)
    case IncludeItem.Eslint:
      return eslintEnabled(dir, ctx)
    case IncludeItem.Githooks:
      return githooksEnabled(dir, ctx)
    case IncludeItem.Jest:
      return jestEnabled(dir, ctx)
    case IncludeItem.LintStaged:
      return lintStagedEnabled(dir, ctx)
    case IncludeItem.Prettier:
      return prettierEnabled(dir, ctx)
    case IncludeItem.Renovate:
      return renovateEnabled(dir, ctx)
  }
}

// https://github.com/conventional-changelog/commitlint#config
const commitlintEnabled = configExists(
  'commitlint',
  [
    'commitlint.config.js',
    '.commitlintrc.js',
    '.commitlintrc.json',
    '.commitlintrc.yml',
  ],
  ['commitlint'],
)

// https://eslint.org/docs/user-guide/configuring#configuration-file-formats
const eslintEnabled = configExists(
  'eslint',
  ['.eslintrc.js', '.eslintrc.yaml', '.eslintrc.yml', '.eslintrc.json'],
  ['eslintConfig'],
)

const githooksEnabled = configExists('git-hook-pure', [], [])

// https://jestjs.io/docs/en/configuration
const jestEnabled = configExists('jest', ['jest.config.js'], [])

// https://github.com/okonet/lint-staged#configuration
const lintStagedEnabled = configExists(
  'lint-staged',
  [
    'lint-staged.config.js',
    '.lintstagedrc',
    '.lintstagedrc.js',
    '.lintstagedrc.json',
    '.lintstagedrc.yml',
    '.lintstagedrc.yaml',
  ],
  ['lint-staged'],
)

const renovateEnabled = configExists(undefined, ['renovate.json'], [])

// https://prettier.io/docs/en/configuration.html
const prettierEnabled = configExists(
  'prettier',
  [
    '.prettierrc',
    '.prettierrc.js',
    '.prettierrc.json',
    '.prettierrc.yml',
    '.prettierrc.yaml',
    '.prettierrc.toml',
    'prettier.config.js',
  ],
  ['prettier'],
)

function configExists(
  packageName: undefined | string,
  fileNames: string[],
  pkgInfoKeys: string[],
): (dir: DirEntry, ctx: TypedSchematicContext<{}, {}>) => boolean {
  return (dir, ctx) => {
    let pkgContent: Record<string, any>
    try {
      pkgContent = JSON.parse(
        dir.file('package.json' as any)!.content.toString(),
      )
    } catch {
      return false
    }

    if (packageName) {
      return Boolean(
        (pkgContent['dependencies'] || {})[packageName] ||
          (pkgContent['devDependencies'] || {})[packageName],
      )
    }

    if (fileNames.some(f => dir.subfiles.includes(f as any))) return true
    return pkgInfoKeys.some(k => pkgContent[k])
  }
}
