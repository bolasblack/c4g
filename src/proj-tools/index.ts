import {
  Rule,
  mergeWith,
  chain,
  callRule,
  apply,
  move,
} from '@angular-devkit/schematics'
import { Observable, merge, of, empty, from } from 'rxjs'
import * as inquirer from 'inquirer'
import {
  installNodePackage,
  NodePackageType,
} from '../utils/rules/installNodePackage'
import { Schema as Options, IncludeItem } from './schema'
import { map, mergeMap, reduce } from 'rxjs/operators'
import { Options as JestInstallOptions } from '../jest-install'
import { file as fileSource } from '../utils/sources/file'
import { schematic } from '../utils/rules/schematic'

export { Options, IncludeItem }

export function main(options: Options): Rule {
  return (tree, ctx) =>
    (options.interactive ? requestOptions(options) : of(options)).pipe(
      map(opts => ({ ...opts, include: opts.include || [] })),
      mergeMap(opts => merge(...opts.include.map(getRule.bind(null, options)))),
      reduce((acc, r) => acc.concat(r), [] as Rule[]),
      mergeMap(rules => {
        if (rules.length) {
          rules = rules.concat(
            installNodePackage({
              packageName: '@c4605/toolconfs',
              workingDirectory: options.cwd,
            }),
          )
        }

        return callRule(chain(rules), tree.branch(), ctx)
      }),
    )
}

function getRule(options: Options, includeItem: IncludeItem): Observable<Rule> {
  const install = (opts: Parameters<typeof installNodePackage>[0]) =>
    installNodePackage({
      workingDirectory: options.cwd || '',
      type: NodePackageType.Dev,
      ...opts,
    })

  const file = (name: string): Rule => copyConfigFile(name, options.cwd || '')

  switch (includeItem) {
    case IncludeItem.Commitlint:
      return of(
        chain([
          file('.commitlintrc.js'),
          install({
            packageName: ['commitlint', '@commitlint/config-conventional'],
          }),
        ]),
      )

    case IncludeItem.LintStaged:
      return of(
        chain([
          file('lint-staged.config.js'),
          install({ packageName: 'lint-staged' }),
        ]),
      )

    case IncludeItem.Githooks:
      return of(install({ packageName: 'git-hook-pure' }))

    case IncludeItem.Renovate:
      return of(file('renovate.json'))

    case IncludeItem.Jest:
      return of(
        schematic('jest-install', {
          cwd: options.cwd,
          react: options.jestReact,
        } as JestInstallOptions),
      )
  }

  return empty()
}

const copyConfigFile = (fileName: string, moveTo: string) =>
  mergeWith(apply(fileSource(`./files/${fileName}`), [move(moveTo)]))

export const requestOptions = (
  opts: Partial<Options> = {},
): Observable<Options> =>
  from(
    inquirer.prompt([
      {
        name: 'include',
        type: 'checkbox',
        choices: Object.values(IncludeItem),
        default: opts.include || [],
      },
      {
        name: 'jestReact',
        type: 'confirm',
        message: 'jest add react support?',
        default: false,
        when(ans: Options) {
          return (
            opts.jestReact == null && ans.include!.includes(IncludeItem.Jest)
          )
        },
      },
    ]),
  )
