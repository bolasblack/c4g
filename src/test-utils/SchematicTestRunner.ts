import { Observable, of as observableOf } from 'rxjs'
import { map } from 'rxjs/operators'
import { logging } from '@angular-devkit/core'
import {
  HostTree,
  Tree,
  Collection,
  TypedSchematicContext,
  ExecutionOptions,
  Schematic,
} from '@angular-devkit/schematics'
import {
  UnitTestTree,
  SchematicTestRunner as SchematicTestRunnerBase,
} from '@angular-devkit/schematics/testing'
import { NodeModulesTestEngineHost } from '@angular-devkit/schematics/tools'

export interface CallSchematic<SchematicSchemaT> {
  (
    schematic: Schematic<{}, {}>,
    options: SchematicSchemaT,
    tree: Observable<Tree>,
    parentContext: Partial<TypedSchematicContext<{}, {}>>,
  ): Observable<Tree>
}

export class SchematicTestRunner extends SchematicTestRunnerBase {
  createContext(
    schematicName: string,
    parent?: Partial<TypedSchematicContext<{}, {}>>,
    executionOptions?: Partial<ExecutionOptions>,
  ) {
    const _collection: Collection<{}, {}> = this['_collection']
    const _logger: logging.Logger = this['_logger']
    const { engine } = this

    const schematic = _collection.createSchematic(schematicName, true)

    return engine.createContext(
      schematic,
      {
        logger: _logger,
        ...parent,
      },
      executionOptions,
    )
  }

  runSchematicAsync<SchematicSchemaT>(
    schematicName: string,
    options?: SchematicSchemaT,
    tree?: Tree,
    callSchematic?: CallSchematic<SchematicSchemaT>,
  ): Observable<UnitTestTree> {
    const _collection: Collection<{}, {}> = this['_collection']
    const _engineHost: NodeModulesTestEngineHost = this['_engineHost']
    const _logger: logging.Logger = this['_logger']

    const schematic = _collection.createSchematic(schematicName, true)
    const host = observableOf(tree || new HostTree())
    _engineHost.clearTasks()

    callSchematic = callSchematic || this._callSchematic.bind(this)

    return callSchematic!(schematic, options as any, host, {
      logger: _logger,
    }).pipe(map(tree => new UnitTestTree(tree)))
  }

  runSchematic<SchematicSchemaT extends {}>(
    schematicName: string,
    options?: SchematicSchemaT,
    tree?: Tree,
    callSchematic?: CallSchematic<SchematicSchemaT>,
  ): UnitTestTree {
    const _collection: Collection<{}, {}> = this['_collection']
    const _engineHost: NodeModulesTestEngineHost = this['_engineHost']
    const _logger: logging.Logger = this['_logger']

    const schematic = _collection.createSchematic(schematicName, true)

    let result: UnitTestTree | null = null
    let error
    const host = observableOf(tree || new HostTree())
    _engineHost.clearTasks()

    callSchematic = callSchematic || this._callSchematic.bind(this)

    callSchematic!(schematic, options as any, host, {
      logger: _logger,
    }).subscribe(t => (result = new UnitTestTree(t)), e => (error = e))

    if (error) {
      throw error
    }

    if (result === null) {
      throw new Error('Schematic is async, please use runSchematicAsync')
    }

    return result
  }

  protected _callSchematic<SchematicSchemaT extends {}>(
    schematic: Schematic<{}, {}>,
    options: SchematicSchemaT | undefined,
    tree: Observable<Tree>,
    parentContext: Partial<TypedSchematicContext<{}, {}>>,
  ) {
    return schematic.call(options || {}, tree, parentContext)
  }
}
