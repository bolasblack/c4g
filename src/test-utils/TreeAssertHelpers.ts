import { UnitTestTree } from '@angular-devkit/schematics/testing'
import { SchematicTestRunner } from '../test-utils/SchematicTestRunner'

export function assertTreeSnapshot(
  runner: SchematicTestRunner,
  tree: UnitTestTree,
) {
  // @ts-ignore
  expect(tree.files).toMatchSnapshot('file structure')
  tree.files.forEach(file => {
    // @ts-ignore
    expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
  })

  // @ts-ignore
  expect(runner.tasks).toMatchSnapshot('rule tasks')
}
