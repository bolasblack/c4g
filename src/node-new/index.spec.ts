import * as path from 'path'
import { Tree, callRule } from '@angular-devkit/schematics'
import { SchematicTestRunner } from '../test-utils/SchematicTestRunner'
import { assertTreeSnapshot } from '../test-utils/TreeAssertHelpers'
import { Options, main as nodeNewRuleFactory } from './index'
import { ProjToolsTypedSchematicContext, IncludeItem } from '../proj-tools'

const collectionPath = path.join(__dirname, '../collection.json')

describe('node-new', () => {
  it('works', async () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)
    const tree = await runner
      .runSchematicAsync<Options>(
        'node-new',
        { name: 'hello', interactive: false },
        Tree.empty(),
      )
      .toPromise()
    assertTreeSnapshot(runner, tree)
  })

  it('support `author` option', async () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)
    const tree = await runner
      .runSchematicAsync<Options>(
        'node-new',
        {
          name: 'hello',
          author: 'c4605 <bolasblack@gmail.com>',
          interactive: false,
        },
        Tree.empty(),
      )
      .toPromise()
    assertTreeSnapshot(runner, tree)
  })

  it('add spec files if user added proj-tool jest', async () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)

    const tree = await runner
      .runSchematicAsync(
        'node-new',
        {
          name: 'hello',
          author: 'c4605 <bolasblack@gmail.com>',
          interactive: false,
        },
        undefined,
        (schematic, opts, tree, oriCtx) => {
          const ctx: ProjToolsTypedSchematicContext<
            {},
            {}
          > = runner.createContext('node-new', oriCtx)
          ctx.projTools = { includes: [IncludeItem.Jest] }

          return callRule(nodeNewRuleFactory(opts), tree, ctx)
        },
      )
      .toPromise()

    assertTreeSnapshot(runner, tree)
  })
})
