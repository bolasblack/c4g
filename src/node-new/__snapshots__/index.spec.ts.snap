// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node-new support \`author\` option: file content: /hello/README.md 1`] = `
"# hello

## Usage

\`\`\`bash
yarn build # build code
yarn watch # build and watch code
\`\`\`
"
`;

exports[`node-new support \`author\` option: file content: /hello/package.json 1`] = `
"{
  \\"name\\": \\"hello\\",
  \\"version\\": \\"0.1.0\\",
  \\"main\\": \\"lib/index.js\\",
  \\"author\\": \\"c4605 <bolasblack@gmail.com>\\",
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"prepublish\\": \\"yarn build\\",
    \\"build\\": \\"tsc --build .\\",
    \\"watch\\": \\"tsc --watch -p .\\"
  },
  \\"devDependencies\\": {},
  \\"dependencies\\": {}
}
"
`;

exports[`node-new support \`author\` option: file content: /hello/src/index.ts 1`] = `
"export function main() {
  console.log('hello from hello')
}

if (require.main === module) {
  main()
}
"
`;

exports[`node-new support \`author\` option: file content: /hello/tsconfig.json 1`] = `
"{
  \\"extends\\": \\"@c4605/toolconfs/tsconfig-node\\",
  \\"compilerOptions\\": {
    \\"outDir\\": \\"lib\\"
  },
  \\"exclude\\": [\\"./lib\\"]
}
"
`;

exports[`node-new support \`author\` option: file content: /hello/tslint.json 1`] = `
"{
  \\"extends\\": \\"@c4605/toolconfs/tslint\\"
}
"
`;

exports[`node-new support \`author\` option: file structure 1`] = `
Array [
  "/hello/README.md",
  "/hello/package.json",
  "/hello/tsconfig.json",
  "/hello/tslint.json",
  "/hello/src/index.ts",
]
`;

exports[`node-new support \`author\` option: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/node-package-install-task",
    "options": Object {
      "packageNames": Array [
        "@c4605/toolconfs",
      ],
      "type": "default",
      "workingDirectory": "hello",
    },
  },
  Object {
    "name": "@c4g/node-package-install-task",
    "options": Object {
      "packageNames": Array [
        "typescript",
        "@types/node",
      ],
      "type": "dev",
      "workingDirectory": "hello",
    },
  },
]
`;

exports[`node-new works: file content: /hello/README.md 1`] = `
"# hello

## Usage

\`\`\`bash
yarn build # build code
yarn watch # build and watch code
\`\`\`
"
`;

exports[`node-new works: file content: /hello/package.json 1`] = `
"{
  \\"name\\": \\"hello\\",
  \\"version\\": \\"0.1.0\\",
  \\"main\\": \\"lib/index.js\\",
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"prepublish\\": \\"yarn build\\",
    \\"build\\": \\"tsc --build .\\",
    \\"watch\\": \\"tsc --watch -p .\\"
  },
  \\"devDependencies\\": {},
  \\"dependencies\\": {}
}
"
`;

exports[`node-new works: file content: /hello/src/index.ts 1`] = `
"export function main() {
  console.log('hello from hello')
}

if (require.main === module) {
  main()
}
"
`;

exports[`node-new works: file content: /hello/tsconfig.json 1`] = `
"{
  \\"extends\\": \\"@c4605/toolconfs/tsconfig-node\\",
  \\"compilerOptions\\": {
    \\"outDir\\": \\"lib\\"
  },
  \\"exclude\\": [\\"./lib\\"]
}
"
`;

exports[`node-new works: file content: /hello/tslint.json 1`] = `
"{
  \\"extends\\": \\"@c4605/toolconfs/tslint\\"
}
"
`;

exports[`node-new works: file structure 1`] = `
Array [
  "/hello/README.md",
  "/hello/package.json",
  "/hello/tsconfig.json",
  "/hello/tslint.json",
  "/hello/src/index.ts",
]
`;

exports[`node-new works: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/node-package-install-task",
    "options": Object {
      "packageNames": Array [
        "@c4605/toolconfs",
      ],
      "type": "default",
      "workingDirectory": "hello",
    },
  },
  Object {
    "name": "@c4g/node-package-install-task",
    "options": Object {
      "packageNames": Array [
        "typescript",
        "@types/node",
      ],
      "type": "dev",
      "workingDirectory": "hello",
    },
  },
]
`;
